core/src/main/java/org/elasticsearch/common/settings/AbstractScopedSettings.java

```diff
@@ -0,0 +1,255 @@
+        if (setting != get(setting.getKey())) {
+            throw new IllegalArgumentException("Setting is not registered for key [" + setting.getKey() + "]");
+        }
+        this.settingUpdaters.add(setting.newUpdater(consumer, logger, settings, predicate));
+    }
+
+    /**
+     * Adds a settings consumer that accepts the values for two settings. The consumer if only notified if one or both settings change.
+     * <p>
+     * Note: Only settings registered in {@link SettingsModule} can be changed dynamically.
+     * </p>
+     * This method registers a compound updater that is useful if two settings are depending on each other. The consumer is always provided
+     * with both values even if only one of the two changes.
+     */
+    public synchronized <A, B> void addSettingsUpdateConsumer(Setting<A> a, Setting<B> b, BiConsumer<A, B> consumer) {
+        if (a != get(a.getKey())) {
+            throw new IllegalArgumentException("Setting is not registered for key [" + a.getKey() + "]");
+        }
+        if (b != get(b.getKey())) {
+            throw new IllegalArgumentException("Setting is not registered for key [" + b.getKey() + "]");
+        }
+        this.settingUpdaters.add(Setting.compoundUpdater(consumer, a, b, logger, settings));
+    }
+
+    /**
+     * Adds a settings consumer.
+     * <p>
+     * Note: Only settings registered in {@link org.elasticsearch.cluster.ClusterModule} can be changed dynamically.
+     * </p>
+     */
+    public synchronized <T> void addSettingsUpdateConsumer(Setting<T> setting, Consumer<T> consumer) {
+       addSettingsUpdateConsumer(setting, consumer, (s) -> {});
+    }
+
+    /**
+     * Transactional interface to update settings.
+     * @see Setting
+     */
+    public interface SettingUpdater {
+        /**
+         * Prepares applying the given settings to this updater. All the heavy lifting like parsing and validation
+         * happens in this method. Yet the actual setting should not be changed by this call.
+         * @param settings the settings to apply
+         * @return <code>true</code> if this updater will update a setting on calling {@link #apply()} otherwise <code>false</code>
+         */
+        boolean prepareApply(Settings settings);
+
+        /**
+         * Applies the settings passed to {@link #prepareApply(Settings)}
+         */
+        void apply();
+
+        /**
+         * Rolls back to the state before {@link #prepareApply(Settings)} was called. All internal prepared state is cleared after this call.
+         */
+        void rollback();
+    }
+
+    /**
+     * Returns the {@link Setting} for the given key or <code>null</code> if the setting can not be found.
+     */
+    public Setting get(String key) {
+        Setting<?> setting = keySettings.get(key);
+        if (setting == null) {
+            for (Map.Entry<String, Setting<?>> entry : groupSettings.entrySet()) {
+                if (entry.getValue().match(key)) {
```

Comment:
why do we need this distinction between normal settings and grouped one? It *seems* we only look up settings be the key they are registered with ?